### PR TITLE
Create 2022-04-06.VPN-Kit.md

### PR DESCRIPTION
**Pull request by christinerose**
To be published on Wednesday, April 6th to stay on schedule with the [MirageOS Release Comms Plan](https://hackmd.io/3M7iWVs3Q4-T0bBb0b0maQ)

Might need a new title....

### CODE REVIEW

**Comment by samoht** [comment]
@djs55 are you ok with the current state of you blog post?

File: 2022-04-06.VPN-Kit.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+---
+updated: 2022-04-06
+authors:
+- name: Dave Scott
+  uri: https://github.com/djs55
+subject: 'A MirageOS VPN Kit'
```
Comment (samoht) [inline 2022-04-06.VPN-Kit.md:6]:
```suggestion
subject: 'How MirageOS powers Docker Desktop'
```

File: 2022-04-06.VPN-Kit.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+permalink: 2022-04-06.VPN-Kit
+---
+
+# A MirageOS VPN Kit
```
Comment (samoht) [inline 2022-04-06.VPN-Kit.md:10]:
```suggestion
# MirageOS and Docker
```

File: 2022-04-06.VPN-Kit.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+[Recently, I posted](https://www.docker.com/blog/how-docker-desktop-networking-works-under-the-hood/) about how [VPN Kit](https://github.com/moby/vpnkit), built with MirageOS libraries, powers [Docker Desktop](https://www.docker.com/products/docker-desktop). Docker Desktop enables users to build and share isolated, *containerised* applications on either a Mac or Windows environment. With 15 million monthly users, it's the most popular developer tool on the planet. Hence, **MirageOS networking libraries transparently handle the traffic of millions of containers**, simplifying many developers' experience.
```
Comment (djs55) [inline 2022-04-06.VPN-Kit.md:12]: Maybe "build and share" -> "build, share and run" (the Docker slogan)

Latest figure according to https://www.docker.com/blog/docker-raises-series-c-build-share-run/ is 16 million monthly users

File: 2022-04-06.VPN-Kit.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+Docker initially started as an easy-to-use packaging of Linux kernels isolation primitives, such as [namespaces](https://en.wikipedia.org/wiki/Linux_namespaces) or [cgroups](https://en.wikipedia.org/wiki/Cgroups). Docker made it convenient for developers to encapsulate applications inside containers, protecting the rest of the system from potential bugs, moving them more easily between machines, and sharing them with other developers. Unfortunately, one of the challenges of running Docker on macOS or Windows is that these Linux primitives are unavailable. Docker Desktop goes to great lengths to emulate them without modifying the user experience of running containers. It runs a (light) Linux VM to host the Docker daemon with additional "glue" helpers to connect the Docker client, running on the host, to that VM. While the Linux VM can run Linux containers, it cannot directly access network resources like internal registries for VPNs specifically configured for forwarding traffic on the host.
```
Comment (djs55) [inline 2022-04-06.VPN-Kit.md:14]: maybe "Linux kernels isolation primitives" -> "the Linux kernel's isolation primitives"

File: 2022-04-06.VPN-Kit.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+VPN Kit bridges that gap and circumvents these issues by reversing the MirageOS network stack. It reads raw ethernet frames coming out of the Linux VM and translates them into macOS or Windows high-level (socket) syscalls: it keeps the connection states for every running Docker container in the Linux VMs and converts all their network traffic transparently into host traffic. This is possible because the MirageOS stack is purely functional: [TCP/IP stack](https://github.com/mirage/mirage-tcpip), DNS, DHCP, (TODO: to complete and add links) (TODO: add the graph from the VPN Kit repo)  model the protocol states as effect-free state machines as described in the [MirageOS TLS paper](https://usenix15.nqsb.io/).
```
Comment (djs55) [inline 2022-04-06.VPN-Kit.md:16]: I like the description of "reversing the MirageOS network stack" -- I hadn't thought of explaining it that way!

**Comment by djs55** [comment]
@christinerose may I push a few tweaks directly to your branch? (Assuming the setting is enabled on github)

**Comment by samoht** [comment]
Yes you can push directly to the branch :-)

**Comment by djs55** [comment]
I still need to add a few links at the bottom

**Comment by samoht** [comment]
might also be worth renaming the file?

**Review by djs55 — APPROVED** [review]

**Pull request merged**